### PR TITLE
virt-launcher: adjust overhead calculation for processes

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
This PR effectively addresses a TODO in the memory overhead calculation for virt-launcher.
It replaces an arbitrary overhead value with the observed memory usage of each process running inside virt-launcher.
A test also ensures we catch any future growth.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/vmi_configuration_test.go
Comment (rmohr) [inline tests/vmi_configuration_test.go:2954]: Should we also stress this fedora to try to max out buffer usages of qemu?

**Review by rmohr — COMMENTED** [review]
Looks great, one question.

File: tests/vmi_configuration_test.go
Comment (jean-edouard) [reply tests/vmi_configuration_test.go:2954]: Yes, good idea. What I used for manual testing triggered oops/panics, so I chose not to include it in the functest :)
However, I assume that's a task `stress-ng` can handle correctly, I'll give it a shot!

Comment (rmohr) [reply tests/vmi_configuration_test.go:2954]: I think we still have some migration tests which create stress in the VMs. It did not work too great and I was glad that we were able to reduce these tests, but you should find some tests which use it already.

Comment (jean-edouard) [reply tests/vmi_configuration_test.go:2954]: Yeah, those were trying to slow down migrations by stressing the guest, not the best idea.
In this case we don't want to actually create stress but just touch all RAM, which it should be able to do.
However, no matter what method we use, if all RAM is in-use that'll at least trigger the guests OOM killer and make a mess, unless... we add swap to the guest? Worth a try.

Comment (rmohr) [reply tests/vmi_configuration_test.go:2954]: Maybe as a follow-up?

**Comment by rmohr** [comment]
/lgtm

@vladikr do you mind having a look too?

/cc @vladikr

File: tests/vmi_configuration_test.go
Comment (xpivarc) [reply tests/vmi_configuration_test.go:2954]: I would personally be up for the swap :+1:

**Comment by jean-edouard** [comment]
This PR was created during the CI outage, the labels are not populated and the tests never started. Trying a close-and-reopen to see if it helps.

**Comment by vladikr** [comment]
> @vladikr do you mind having a look too?
> 
> /cc @vladikr

Looking :)

**Comment by jean-edouard** [comment]
/hold
The flake-checker already caught virt-launcher using more RAM than usual.
That lane is a good benchmark for this PR and must pass before we merge.

**Comment by vladikr** [comment]
In general, everything here looks good to me.
However, I'd make this test OOM friendly, by running the VMI with a guaranteed QOS. I'd also run stress. Even with 25% of the total guest memory, it will be efficient. 
If you think running stress is not a good option, then perhaps we can allocate a vfio device, this will lock all the guest memory.

**Comment by jean-edouard** [comment]
> In general, everything here looks good to me. However, I'd make this test OOM friendly, by running the VMI with a guaranteed QOS. I'd also run stress. Even with 25% of the total guest memory, it will be efficient. If you think running stress is not a good option, then perhaps we can allocate a vfio device, this will lock all the guest memory.

Thanks for the review! The topic of stressing the guest was discussed above, and I think we came to an agreement for doing it as a follow-up effort.
The point here was more to account for the various processes running in virt-launcher rather than tackling guest memory usage, which is already handled separately in the code.
I am planning on testing that code on large nodes in the near future, fixing it if necessary and adding a functional test with a stressed guest.

**Comment by vladikr** [comment]
> > In general, everything here looks good to me. However, I'd make this test OOM friendly, by running the VMI with a guaranteed QOS. I'd also run stress. Even with 25% of the total guest memory, it will be efficient. If you think running stress is not a good option, then perhaps we can allocate a vfio device, this will lock all the guest memory.
> 
> Thanks for the review! The topic of stressing the guest was discussed above, and I think we came to an agreement for doing it as a follow-up effort. The point here was more to account for the various processes running in virt-launcher rather than tackling guest memory usage, which is already handled separately in the code. I am planning on testing that code on large nodes in the near future, fixing it if necessary and adding a functional test with a stressed guest.

Alright, I'm fine with this as long as we have a plan for a follow-up. 

/approve

**Review by acardace — CHANGES_REQUESTED** [review]
Looks good, just 2 nits.

Also, can you improve the commit message by pretty much doing a copy-paste of the PR description? Otherwise, we'd lose what you wrote there which helps to understand the reason for this patch.

File: pkg/virt-controller/services/template.go
Comment (iholder-redhat) [inline pkg/virt-controller/services/template.go:130]: Do these need to be exposed?

Comment (jean-edouard) [reply pkg/virt-controller/services/template.go:130]: Exposed as in exported/uppercase? They do but only for the functional test.

Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go:130]: Yes, that's what I meant :)

Got you. Well, too bad golang doesn't have a way of exporting variables only to functional tests. I don't have any ideas expect for duplicating the variables, but I don't think it's better.

Alright, thanks!

**Comment by jean-edouard** [comment]
/hold
Needs to be adjusted for https://github.com/kubevirt/kubevirt/pull/7451

File: tests/vmi_configuration_test.go
Comment (jean-edouard) [reply tests/vmi_configuration_test.go]: Thank you @acardace for your review. I somehow missed it!
Thanks for the tip here, I fixed this and the other one.

File: pkg/virt-controller/services/template.go
Comment (jean-edouard) [reply pkg/virt-controller/services/template.go:127]: `VirtLauncherOverhead` was the sum of the memory consumption of the 2 virt-launcher processes, 2*75=150Mi.
https://github.com/kubevirt/kubevirt/pull/7451 introduced virt-launcher-monitor, which replaces one of the 2 virt-launcher processes and uses 50Mi less than virt-launcher, so 25Mi.

Comment (xpivarc) [reply pkg/virt-controller/services/template.go:127]: I was comparing the values to https://github.com/kubevirt/kubevirt/commit/2c8bdb6ad9fa065bd66977b5c0e83f4966741e7d. I just wanted to make sure.

Comment (jean-edouard) [reply pkg/virt-controller/services/template.go:127]: Ah, yes, these were the initial values for this PR. They were underestimated on purpose, to have CI trip on them and give me the right numbers.

Comment (xpivarc) [reply pkg/virt-controller/services/template.go:127]: Thanks for the explanation!

**Review by acardace — APPROVED** [review]

**Comment by xpivarc** [comment]
/hold


```
Tests  Suite: [sig-compute]Configurations VirtualMachineInstance definition
[Serial][rfe_id:2065][crit:medium][vendor:cnv-qe@redhat.com][level:component]with  3 CPU cores [test_id:1659]should report 3 cpu cores under guest OS expand_less | 16s
-- | --
{ 
Failure tests/vmi_configuration_test.go:241 
Expected     <int64>: 339 to equal 
    <int64>: 303 
tests/vmi_configuration_test.go:279}


```

**Comment by vladikr** [comment]
@jean-edouard `tests/vmi_configuration_test.go:2949:11: undefined: libnet`

**Pull request merged**
